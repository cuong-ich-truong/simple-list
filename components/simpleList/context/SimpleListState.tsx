export interface BaseItem {
  id: string;
  name: string;
  price: number;
}
export interface LineItem extends BaseItem {
  categoryId: string;
}

export interface Category extends BaseItem {
  lineItems: LineItem[];
}

export default interface SimpleListState {
  categories: Category[];
  total: number;
}
