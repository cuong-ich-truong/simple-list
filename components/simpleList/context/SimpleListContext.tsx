import React from 'react';
import SimpleListState from './SimpleListState';

export default React.createContext<SimpleListState>({
  categories: [],
  total: 0
});
