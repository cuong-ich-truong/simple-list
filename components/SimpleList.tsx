import React, { useContext } from 'react';
import SimpleListContext from './simpleList/context/SimpleListContext';
import LineItem from './simpleList/LineItem';
import StHeader from './simpleList/StHeader';
import StFooter from './simpleList/StFooter';
import StItem from './simpleList/StItem';
import LineItemForm from './simpleList/LineItemForm';

const SimpleList: React.FC = () => {
  const { categories, total } = useContext(SimpleListContext);

  return (
    <>
      {categories.map(({ id: categoryId, name, price, lineItems }) => (
        <div key={categoryId}>
          <StHeader>
            <LineItem name={name} price={price} />
          </StHeader>
          <StItem>
            {lineItems.map(({ id: subItemId, name, price }) => (
              <LineItem key={subItemId} name={name} price={price} />
            ))}
          </StItem>
        </div>
      ))}
      <StFooter>
        <LineItem name="TOTAL" price={total} />
      </StFooter>
      <LineItemForm />
    </>
  );
};

export default SimpleList;
