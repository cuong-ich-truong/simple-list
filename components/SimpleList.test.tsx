import { render } from '@testing-library/react';
import SimpleList from './SimpleList';
import SimpleListContext from './simpleList/context/SimpleListContext';
import SimpleListState, {
  Category
} from './simpleList/context/SimpleListState';

const MockProvider: React.FC<Partial<SimpleListState>> = ({
  categories = [],
  total = 0,
  addLineItem = jest.fn(),
  removeLineItem = jest.fn(),
  children
}) => {
  return (
    <SimpleListContext.Provider
      value={{ categories, total, addLineItem, removeLineItem }}
    >
      {children}
    </SimpleListContext.Provider>
  );
};

describe('Home', () => {
  let container: HTMLElement;
  let asFragment: () => DocumentFragment;
  let mockCategories: Category[];

  beforeEach(() => {
    mockCategories = [
      {
        id: '6c8f51aa-9361-48b7-8600-aaa1cf60dd8e',
        name: '81df986b-a2ce-4b70-b4b2-12e3a41bf13a',
        price: 0,
        lineItems: []
      },
      {
        id: 'e350f15e-343c-45d6-a314-ee5984f3482b',
        name: '1754cdd9-7a6f-4f48-9816-4ee29949caa5',
        price: 0,
        lineItems: []
      }
    ];

    ({ container, asFragment } = render(
      <MockProvider categories={mockCategories}>
        <SimpleList />
      </MockProvider>
    ));
  });

  it('should render without crashing', () => {
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render category names', () => {
    expect(container).toHaveTextContent(mockCategories[0].name);
    expect(container).toHaveTextContent(mockCategories[1].name);
  });
});
