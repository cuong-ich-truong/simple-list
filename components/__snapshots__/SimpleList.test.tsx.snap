// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should render without crashing 1`] = `
<DocumentFragment>
  <div>
    <div
      class="StHeader-sc-9xi8n7-0 dOWrXh"
    >
      <div
        class="StLine-dhselc-0 bPoNII"
      >
        <div
          class="StTitle-cua8m8-0 bhCPJC"
        >
          F7tkS
        </div>
        <div
          class="StPrice-cijbu8-0 klpCDV"
        >
          <div>
            $0.00
          </div>
        </div>
      </div>
    </div>
    <div
      class="StItem-sc-1ncxcux-0 kWpKVU"
    />
  </div>
  <div>
    <div
      class="StHeader-sc-9xi8n7-0 dOWrXh"
    >
      <div
        class="StLine-dhselc-0 bPoNII"
      >
        <div
          class="StTitle-cua8m8-0 bhCPJC"
        >
          HDNMgnb1NmjVbo
        </div>
        <div
          class="StPrice-cijbu8-0 klpCDV"
        >
          <div>
            $0.00
          </div>
        </div>
      </div>
    </div>
    <div
      class="StItem-sc-1ncxcux-0 kWpKVU"
    />
  </div>
  <div>
    <div
      class="StHeader-sc-9xi8n7-0 dOWrXh"
    >
      <div
        class="StLine-dhselc-0 bPoNII"
      >
        <div
          class="StTitle-cua8m8-0 bhCPJC"
        >
          IodB7fN61uLmYY&t#w
        </div>
        <div
          class="StPrice-cijbu8-0 klpCDV"
        >
          <div>
            $0.00
          </div>
        </div>
      </div>
    </div>
    <div
      class="StItem-sc-1ncxcux-0 kWpKVU"
    />
  </div>
  <div
    class="StFooter__StHeader-sc-10ahsio-0 kIxEMy"
  >
    <div
      class="StLine-dhselc-0 bPoNII"
    >
      <div
        class="StTitle-cua8m8-0 bhCPJC"
      >
        TOTAL
      </div>
      <div
        class="StPrice-cijbu8-0 klpCDV"
      >
        <div>
          $0.00
        </div>
      </div>
    </div>
  </div>
  <div
    class="StLine-dhselc-0 jhSrAX"
  >
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled"
        data-shrink="false"
        for="item-name-input"
        id="item-name-input-label"
      >
        Item Name
      </label>
      <div
        class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-fullWidth MuiInputBase-formControl"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiFilledInput-input"
          id="item-name-input"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-filled"
        data-shrink="true"
        for="item-price-input"
      >
        Price
      </label>
      <div
        class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl MuiInputBase-adornedStart MuiFilledInput-adornedStart"
      >
        <div
          class="MuiInputAdornment-root MuiInputAdornment-filled MuiInputAdornment-positionStart"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            $
          </p>
        </div>
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputAdornedStart MuiFilledInput-inputAdornedStart"
          id="item-price-input"
          min="0"
          step="1"
          type="number"
          value="0"
        />
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-filled Mui-required Mui-required"
        data-shrink="true"
        for="category-select-label"
      >
        Category
        <span
          aria-hidden="true"
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
        >
          â€‰*
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="category-select-label item-category-select"
          class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-filled MuiInputBase-input MuiFilledInput-input"
          id="item-category-select"
          role="button"
          tabindex="0"
        >
          F7tkS
        </div>
        <input
          aria-hidden="true"
          class="MuiSelect-nativeInput"
          required=""
          tabindex="-1"
          value="59aa9abd-8825-5b65-b3ba-1adb09352678"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconFilled"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Add
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
